File: Design-Patterns-Using-Cpp/Adapter/AdapterPattern.cpp
#include <iostream>

// Desired Interface
class DesiredInterface {
public:
    virtual void operation() = 0;
    virtual ~DesiredInterface() {} // Virtual destructor for proper cleanup
};

// Adaptee class
class Adaptee {
public:
    void someOperation() {
        std::cout << "Adaptee someOperation() function called." << std::endl;
    }
};

// Adapter class
class Adapter : public DesiredInterface {
private:
    Adaptee adaptee;

public:
    void operation() override {
        adaptee.someOperation();
    }
};

// Client Code
int main() {
    DesiredInterface* adapter = new Adapter();
    adapter->operation();

    // Clean up memory
    delete adapter;

    return 0;
}

/*
Adaptee someOperation() function called.
*/

File: Design-Patterns-Using-CSharp/Adapter/AdapterPattern.cs
using System;

// Desired Interface
interface DesiredInterface
{
    void Operation();
}

// Adapter class
class Adapter : DesiredInterface
{
    private Adaptee adaptee;

    public Adapter()
    {
        this.adaptee = new Adaptee();
    }

    public void Operation()
    {
        adaptee.SomeOperation();
    }
}

// Adaptee class
class Adaptee
{
    public void SomeOperation()
    {
        Console.WriteLine("Adaptee SomeOperation() function called.");
    }
}

// Client Code
public class AdapterPattern
{
    public static void Main(string[] args)
    {
        DesiredInterface adapter = new Adapter();
        adapter.Operation();
    }
}
/* 
Adaptee someOperation() function called.
*/